exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java

```diff
@@ -208,37 +209,70 @@ public void initiateAddingService(Fork fork, ServiceInstanceSpec instanceSpec,
+   * by {@link #initiateAddingService(Fork, ServiceInstanceSpec, byte[])} in advance.
+   * Activation leads to the service instance registration, allowing subsequent operations on it:
    * transactions, API requests.
+   * Stopping leads to the service disabling i.e. stopped service does not execute transactions,
+   * process events, provide APIs, etc. And it becomes unavailable to other services,
+   * but still exists.
    *
    * @param instanceSpec a service instance specification; must reference a deployed artifact
    * @param instanceStatus a new status of the service instance
-   * @throws IllegalArgumentException if the service is already started; or its artifact
-   *     is not deployed
+   * @throws IllegalArgumentException if activating already active service; or its artifact
+   *     is not deployed; or unrecognized service status received
    */
   public void updateInstanceStatus(ServiceInstanceSpec instanceSpec,
       InstanceState.Status instanceStatus) {
-    try {
-      synchronized (lock) {
-        // Create a previously added service
-        ServiceWrapper service = createService(instanceSpec);
-        // Register it in the runtime
-        registerService(service);
-        // Connect its API
-        connectServiceApi(service);
+    synchronized (lock) {
+      if (instanceStatus.equals(Status.ACTIVE)) {
+        activateService(instanceSpec);
+      } else if (instanceStatus.equals(Status.STOPPED)) {
+        stopService(instanceSpec);
+      } else if (instanceStatus.equals(Status.NONE)) {
+        logger.warn("None status for the service instance {}, ignore. "
+            + "Possibly restoring services state after reboot?", instanceSpec.getName());
+      } else {
+        String msg = String.format("Unexpected status %s received for the service %s",
+            instanceStatus.name(), instanceSpec.getName());
+        logger.error(msg);
+        throw new IllegalArgumentException(msg);
       }
+    }
+  }
 
-      logger.info("Added a service: {}", instanceSpec);
+  private void activateService(ServiceInstanceSpec instanceSpec) {
+    try {
+      // Create a previously added service
+      ServiceWrapper service = createServiceInstance(instanceSpec);
+      // Register it in the runtime
+      registerService(service);
+      // Connect its API
+      connectServiceApi(service);
+      logger.info("Activated a service: {}", instanceSpec);
     } catch (Exception e) {
-      logger.error("Failed to add a service {} instance", instanceSpec, e);
+      logger.error("Failed to activate a service {} instance", instanceSpec, e);
       throw e;
     }
   }
 
-  private ServiceWrapper createService(ServiceInstanceSpec instanceSpec) {
+  private void stopService(ServiceInstanceSpec instanceSpec) {
+    // TODO: ECR-2334 add restriction for creation new snapshots
+    String name = instanceSpec.getName();
+    Optional<ServiceWrapper> activeService = findService(name);
```

Comment:
Why does this code perform a no-op when stopping of a non-existent service
is requested? Shan't it just throw? In this case, a simpler

```java
int id = instanceSpec.getId();
ServiceWrapper service = getServiceById(id); // Will throw if does not exist.
```

will do.

Reply:
we will have non-existent service in case of restoring state

Reply:
Interesting — I'd consider if any changes to how core works are needed, if we don't miss anything: https://github.com/exonum/exonum-java-binding/pull/1358#discussion_r367506172